lisp/bazel-build.el

```diff
@@ -25,38 +25,109 @@
 ;;; Code:
 
 (require 'bazel-util)
+(require 'subr-x)
 
 (defun bazel-build (target)
   "Build a Bazel TARGET."
-  (interactive (list (bazel-build--read-target "bazel build ")))
+  (interactive (list (bazel-build--read-target "bazel build")))
   (bazel-build--run-bazel-command "build" target))
 
 (defun bazel-run (target)
   "Build and run a Bazel TARGET."
-  (interactive (list (bazel-build--read-target "bazel run ")))
+  (interactive (list (bazel-build--read-target "bazel run")))
   (bazel-build--run-bazel-command "run" target))
 
 (defun bazel-test (target)
   "Build and run a Bazel test TARGET."
-  (interactive (list (bazel-build--read-target "bazel test ")))
+  (interactive (list (bazel-build--read-target "bazel test")))
   (bazel-build--run-bazel-command "test" target))
 
+(defvar bazel-command "bazel"
+  "Bazel executable to run for subprocesses.")
+
+(defvar bazel-query-args '("query" "--noblock_for_lock")
+  "Bazel query subcommand and arguments.")
+
+(defvar bazel-run-commands-from-root nil
+  "A flag indicating to run commands from the workspace root.")
+
 (defun bazel-build--run-bazel-command (command target)
   "Run Bazel tool with given COMMAND, e.g. build or run, on the given TARGET."
-  (compile
-   (mapconcat #'shell-quote-argument (list "bazel" command target) " ")))
-
-(defun bazel-build--read-target (prompt)
-  "Read a Bazel build target from the minibuffer.  PROMPT is a read-only prompt."
-  (let* ((file-name (buffer-file-name))
-         (workspace-root
-          (or (bazel-util-workspace-root file-name)
-              (user-error "Not in a Bazel workspace.  No WORKSPACE file found")))
-         (package-name
-          (or (bazel-util-package-name file-name workspace-root)
-              (user-error "Not in a Bazel package.  No BUILD file found")))
-         (initial-input (concat "//" package-name)))
-    (read-string prompt initial-input)))
+  (let ((command
+         (if bazel-run-commands-from-root
+             (let ((root-dir (or (bazel-util-workspace-root
+                                  (or (buffer-file-name)
+                                      (and dired-directory (directory-file-name dired-directory))))
+                                 (user-error "Could not find workspace."))))
+               (format "cd %s && %s %s %s" root-dir
+                       bazel-command command target))
+           (mapconcat #'shell-quote-argument (list bazel-command command target) " "))))
+    (compile command)))
+
+(defun bazel-build--process (&rest args)
```

Comment:
Right now this function is only used to query a target list for completion, right? Then I think signaling an error and popping up the error message is too drastic, since not being able to do completion isn't a very critical failure mode. I'd suggest just not performing completion in that case and fall back to the previous package-only method.